<a href="https://colab.research.google.com/github/d-melamed/DS-Unit-1-Sprint-2-Statistics/blob/master/module1/DM_LS_DS_121_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

## Practice 1-Sample T-tests

One Sample t-tests determine whether or not a sample mean is statistically different from some known (or hypothesized) population mean. 

### 1) Load the Data
- Use the [automobile dataset](https://archive.ics.uci.edu/ml/datasets/Automobile)
- Fix the column headers
- Make sure NaNs are used to indicate missing values

Feel free to add code cells and text cells as needed throughout the assignment.

In [4]:
### YOUR WORK HERE

import pandas as pd
import numpy as np

column_headers = ['symboling', 'normalized-losses', 'make', 'fuel-type', 'aspiration', 'num-of-doors', 'body-style', 'drive-wheels', 'engine-location', 'wheel-base', 'length', 'width', 'height', 'curb-weight', 'engine-type', 'num-of-cylinders', 'engine-size', 'fuel-system', 'bore', 'stroke', 'compression-ratio', 'horsepower', 'peak-rpm', 'city-mpg', 'highway-mpg', 'price']
df = pd.read_csv("/content/imports-85.data", names=column_headers)

In [5]:
df.head()

symboling normalized-losses         make  ... city-mpg highway-mpg  price
0          3                 ?  alfa-romero  ...       21          27  13495
1          3                 ?  alfa-romero  ...       21          27  16500
2          1                 ?  alfa-romero  ...       19          26  16500
3          2               164         audi  ...       24          30  13950
4          2               164         audi  ...       18          22  17450

[5 rows x 26 columns]

In [6]:
df = df.replace({'?':np.NaN})
df.head()

symboling normalized-losses         make  ... city-mpg highway-mpg  price
0          3               NaN  alfa-romero  ...       21          27  13495
1          3               NaN  alfa-romero  ...       21          27  16500
2          1               NaN  alfa-romero  ...       19          26  16500
3          2               164         audi  ...       24          30  13950
4          2               164         audi  ...       18          22  17450

[5 rows x 26 columns]

### 2) Pretend that this dataset represents the cars at a used car lot in your local town. 

- Use df.sample() to pick a random sample of 10 cars. Note that because this sample is random we are going to set the `random_state` so that all of us in the class get the same random sample. Please set your random state to `30` when using `df.sample()`

Is your sample reflective of the population value in regards to highway-mpg? Find the mean for `highway-mpg` for the entire dataset and use a 1-sample t-test to compare your estimated sample mean to the population mean. Can you say that they are different? 

In [7]:
### YOUR WORK HERE
sample = df.sample(10, random_state=30)

sample.head()

symboling normalized-losses           make  ... city-mpg highway-mpg  price
121          1               154       plymouth  ...       31          38   6692
176         -1                65         toyota  ...       27          32  10898
24           1               148          dodge  ...       31          38   6229
88          -1               137     mitsubishi  ...       23          30   9279
72           3               142  mercedes-benz  ...       16          18  35056

[5 rows x 26 columns]

In [8]:
df['highway-mpg'].mean()

30.75121951219512

In [9]:
sample['highway-mpg'].mean()

32.8

### Null Hypothesis:

Sample mean is equal to 30.75, i.e., sample is reflective of all cars on lot.

### Alternative Hypothesis:

Sample mean is not equal to 30.75, i.e., sample is not reflective of all cars on lot.

### Confidence Level: 90%

In [10]:
from scipy import stats

In [11]:
stats.ttest_1samp(sample['highway-mpg'], df['highway-mpg'].mean(), nan_policy='omit')

Ttest_1sampResult(statistic=0.8912422391619923, pvalue=0.39600040771093326)

### Conclusion:
Based on a t-statistic of .891 and a p-value of .396, ***we fail to reject*** the null hypothesis that the sample is reflective of all cars on the lot.

The salesman says the cars he sells typically have a fuel efficiency of about 35 miles per gallon on the highway. You want to verify his claim but can only test 10 cars. Using your sample of 10, test his claim and report your results.

In [12]:
### YOUR WORK HERE
stats.ttest_1samp(sample['highway-mpg'], 35, nan_policy='omit')

Ttest_1sampResult(statistic=-0.9570244044334747, pvalue=0.36355472283248624)

### Conclusion:

Based on a t-statistic of -.957 and a p-value of .364, ***we fail to reject*** the null hypothesis that the salesman's reported fuel efficiency is the same as the sample.

Imagine that you now have the capacity to test 100 cars on the lot. Using the same random state of `30`, take a sample of 100 cars. Run a t-test to verify the salesman's claim with your new larger sample. Do you reach the same conclusion as you did with the sample size of 10? 

In [13]:
### YOUR WORK HERE
big_sample = df.sample(100, random_state=30)

In [14]:
big_sample['highway-mpg'].mean()

30.64

In [15]:
stats.ttest_1samp(big_sample['highway-mpg'], 35, nan_policy='omit')

Ttest_1sampResult(statistic=-6.5185242467032305, pvalue=3.003137417468754e-09)

### Conclusion:

Based on a t-statistic of -6.52 and a p-value of .000000003, we ***reject*** the null hypothesis that the salesman's claim is true and the cars get 35 highway-mpg. The sample is different (aka he's a liar)

Why might these two t-tests using the same dataset lead to different conclusions about the salesman's claim?

***Larger sample size leads to a more accurate depiction of the population. The smaller sample size has a higher chance of being biased.***

## Practice 2-Sample T-tests

Two Sample t-tests determine whether or not two sample means are statistically different from each other. 

This portion of your assignment requires you to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). 


### 1) Load the data

The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

- Read the dataset in from UCI, you'll need to provide a list of column headers
- Encode "yes" votes as 1 and "no" votes as 0. (You can use `df.replace()` to swap out these values)
- Use dataframe filtering to split the dataframe into two new dataframes based on party. Hold all republicans in one dataframe and all democrats in the other. These will be our two different "samples."



In [16]:
### YOUR WORK HERE
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data

--2020-09-05 01:44:38--  https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18171 (18K) [application/x-httpd-php]
Saving to: ‘house-votes-84.data’

house-votes-84.data 100%[===================>]  17.75K  --.-KB/s    in 0.07s   

2020-09-05 01:44:39 (271 KB/s) - ‘house-votes-84.data’ saved [18171/18171]



In [17]:
column_headers = ['party','handicapped-infants','water-project',
                  'budget','physician-fee-freeze', 'el-salvador-aid',
                  'religious-groups','anti-satellite-ban',
                  'aid-to-contras','mx-missile','immigration',
                  'synfuels', 'education', 'right-to-sue','crime','duty-free',
                  'south-africa']

df = pd.read_csv('house-votes-84.data', names=column_headers)

print(df.shape)
df.head()

(435, 17)


party handicapped-infants water-project  ... crime duty-free south-africa
0  republican                   n             y  ...     y         n            y
1  republican                   n             y  ...     y         n            ?
2    democrat                   ?             y  ...     y         n            n
3    democrat                   n             y  ...     n         n            y
4    democrat                   y             y  ...     y         y            y

[5 rows x 17 columns]

In [18]:
df = df.replace({'y':1, 'n':0, "?":np.NaN})

df.head()

party  handicapped-infants  ...  duty-free  south-africa
0  republican                  0.0  ...        0.0           1.0
1  republican                  0.0  ...        0.0           NaN
2    democrat                  NaN  ...        0.0           0.0
3    democrat                  0.0  ...        0.0           1.0
4    democrat                  1.0  ...        1.0           1.0

[5 rows x 17 columns]

In [19]:
rep = df[df['party'] =='republican']
dem = df[df['party'] =='democrat']

### 2) Perform two sample T-tests on different issues and report the results.

- Find an issue that democrats support more than republicans with p < 0.01 (significant at the 99% level).
- Find an issue that republicans support more than democrats with p < 0.01 (significant at the 99% level).
- Find an issue where the difference between republicans and democrats has p > 0.1 (Not significant at the 90% level - i.e. there may not be much of a difference the two sample means)

Please for each test that you run state your null and alternative hypothesis and then write a conclusion reflecting on the null and alternative hypothesis.

Remember, that two-sample t-tests will only tell us if the two groups are *different* from one another. We'll have to look at their sample means directly or use the sign on the t-statistic to know which of the two sample means is larger. 

In [20]:
### YOUR WORK HERE

In [21]:
df.mean()

handicapped-infants     0.442080
water-project           0.503876
budget                  0.596698
physician-fee-freeze    0.417453
el-salvador-aid         0.504762
religious-groups        0.641509
anti-satellite-ban      0.567696
aid-to-contras          0.576190
mx-missile              0.501211
immigration             0.504673
synfuels                0.362319
education               0.423267
right-to-sue            0.509756
crime                   0.593301
duty-free               0.427518
south-africa            0.812689
dtype: float64

### Null Hypothesis:

Democrat support for the `budget` bill is equal to Republican levels of support from the `budget` bill.

### Alternative Hypothesis:

Democrat support for the `budget` bill is not equal to Republican levels of support from the `budget` bill.

### Confidence level: 99%

In [22]:
stats.ttest_ind(dem['budget'],rep['budget'], nan_policy='omit')

Ttest_indResult(statistic=23.21277691701378, pvalue=2.0703402795404463e-77)

###Conclusion:

Based on a t-stat of 23.2 and a p-value of 2.07e-77, we ***reject*** the null hypothesis that Democrat levels of support are the same as Republican levels on the  `budget` bill. Democrats supported the bill more (t-stat is positive, left-hand sample mean is larger).

### Null Hypothesis:

Republican support for the `religious-groups` bill is equal to democrat levels of support from the `religious-groups` bill.

### Alternative Hypothesis:

Republican support for the `religious-groups` bill is not equal to democrat levels of support from the `religious-groups` bill.

### Confidence level: 99%

In [23]:
stats.ttest_ind(dem['religious-groups'],rep['religious-groups'], nan_policy='omit')

Ttest_indResult(statistic=-9.737575825219457, pvalue=2.3936722520597287e-20)

###Conclusion:

Based on a t-stat of -9.74 and a p-value of 2.39e-20, we ***reject*** the null hypothesis that Republican levels of support are the same as Democrat levels on `religious-groups` bill. Republicans supported the bill more (t-stat is negative, right-hand sample mean is larger).

### Null Hypothesis:

Democrat support for the `water-project` bill is equal to Republican levels of support from the `water-project` bill.

### Alternative Hypothesis:

Democrat support for the `water-project` bill is not equal to Republican levels of support from the `water-project` bill.

### Confidence level: 90%

In [39]:
stats.ttest_ind(dem['water-project'],rep['water-project'], nan_policy='omit')

Ttest_indResult(statistic=-0.08896538137868286, pvalue=0.9291556823993485)

### Conclusion:

Based on a t-stat of -.089 and a pvalue of .93, we ***fail to reject*** the null hypothesis that the Democrat support for the `water-project` bill is the same as the Republican support.

## Stretch Goals:

### 1) Use functions and some iterator (for loop, .apply(), list comprehension, etc.) to perform two sample t-tests on every issue in the dataset in an automated fashion.

In [25]:
### YOUR WORK HERE

def sample_ts ()



### 2) Work on Performing a T-test without using Scipy in order to get "under the hood" and learn more thoroughly about this topic.
### Start with a 1-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://blog.minitab.com/hs-fs/hubfs/Imported_Blog_Media/701f9c0efa98a38fb397f3c3ec459b66.png?width=247&height=172&name=701f9c0efa98a38fb397f3c3ec459b66.png) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

### Be sure to check your work using Scipy!





In [26]:
### YOUR WORK HERE

### 3) Then try a 2-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://lh3.googleusercontent.com/proxy/rJJ5ZOL9ZDvKOOeBihXoZDgfk7uv1YsRzSQ1Tc10RX-r2HrRpRLVqlE9CWX23csYQXcTniFwlBg3H-qR8MKJPBGnjwndqlhDX3JxoDE5Yg) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Be sure to check your work using Scipy!

In [27]:
### YOUR WORK HERE